File: Lesson 27/Lesson 27/DrawingView.m
//
//  DrawingView.m
//  Lesson 27
//
//  Created by Lucas Derraugh on 12/31/11.
//  Copyright (c) 2011 Lucas Derraugh. All rights reserved.
//

#import "DrawingView.h"

@implementation DrawingView {
    NSPoint startPoint;
    NSPoint endPoint;
}

- (void)drawRect:(NSRect)dirtyRect
{
    [NSColor.redColor set];
    NSRectFill(dirtyRect);
    
    [NSColor.blueColor set];
    [[NSBezierPath bezierPathWithRect:NSMakeRect(startPoint.x, startPoint.y, endPoint.x-startPoint.x, endPoint.y-startPoint.y)] stroke];
}

- (void)mouseDown:(NSEvent *)theEvent {
    NSPoint point = theEvent.locationInWindow;
    startPoint = [self convertPoint:point fromView:nil];
}

- (void)mouseDragged:(NSEvent *)theEvent {
    NSPoint point = theEvent.locationInWindow;
    endPoint = [self convertPoint:point fromView:nil];
    [self setNeedsDisplay:YES];
}

@end

File: Lesson 63/Lesson 63/DrawingView.swift
//
//  DrawingView.swift
//  Lesson 63
//
//  Created by Lucas Derraugh on 12/31/14.
//  Copyright (c) 2014 Lucas Derraugh. All rights reserved.
//

import Cocoa

class DrawingView: NSView {
    
    override func mouseDown(with theEvent: NSEvent) {
        let point = convert(theEvent.locationInWindow, from: nil)
        if theEvent.modifierFlags.contains(.option) {
            if let view = hitTest(point), view is CircleView {
                removeCircle(view)
                undoManager?.setActionName("Remove Circle")
            }
        } else {
            let circle = CircleView(frame: NSMakeRect(point.x-15, point.y-15, 30, 30))
            addCircle(circle)
            undoManager?.setActionName("Add Circle")
        }
    }
    
    func addCircle(_ circle: NSView) {
        addSubview(circle)
        undoManager?.registerUndo(withTarget: self, selector: #selector(removeCircle(_:)), object: circle)
    }
    
    func removeCircle(_ circle: NSView) {
        circle.removeFromSuperview()
        (undoManager?.prepare(withInvocationTarget: self) as AnyObject).addCircle(circle)
    }

    override var wantsUpdateLayer: Bool {
        return true
    }
    
    override func updateLayer() {
        layer?.backgroundColor = NSColor.black.cgColor
    }
    
}
